conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java

```diff
@@ -0,0 +1,184 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.undertow.runtime;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.conjure.java.api.errors.SerializableError;
+import com.palantir.conjure.java.serialization.ObjectMappers;
+import com.palantir.conjure.java.undertow.lib.Endpoint;
+import com.palantir.conjure.java.undertow.lib.ReturnValueWriter;
+import com.palantir.conjure.java.undertow.lib.UndertowRuntime;
+import io.undertow.Undertow;
+import io.undertow.server.HttpHandler;
+import io.undertow.util.Headers;
+import io.undertow.util.HttpString;
+import io.undertow.util.Methods;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.OptionalInt;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public final class AsyncRequestProcessingTest {
+
+    private static final ObjectMapper clientMapper = ObjectMappers.newClientObjectMapper();
+
+    private Undertow server;
+    private ListeningScheduledExecutorService scheduler;
+
+    @Before
+    public void before() {
+        this.scheduler = MoreExecutors.listeningDecorator(Executors.newSingleThreadScheduledExecutor());
+        UndertowRuntime runtime = ConjureUndertowRuntime.builder()
+                // Use a 1 second timeout so we can wait for the timeout to be hit without tests running too long.
+                .asyncTimeout(Duration.ofSeconds(1))
+                .build();
+        ReturnValueWriter<String> writer = (returnValue, exchange) -> {
+            exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "text/plain");
+            exchange.getResponseSender().send(returnValue);
+        };
+        HttpHandler httpHandler = exchange -> {
+            OptionalInt maybeDelay = runtime.plainSerDe()
+                    .deserializeOptionalInteger(exchange.getQueryParameters().get("delay"));
+            final ListenableFuture<String> future;
+            if (maybeDelay.isPresent()) {
+                int delay = maybeDelay.getAsInt();
+                future = Futures.transform(
+                        scheduler.schedule(() -> { }, delay, TimeUnit.MILLISECONDS),
```

Comment:
I think we need to help users of the async framework define a sensible executor for their work. maybe a `ConjureAsyncRuntime` could provide such a thing?

Reply:
Our server library already provides an executor factory, and service implementations don't get access to the `ConjureRuntime` and by extension the `ConjureAsyncRuntime`.

We shouldn't provide executors to services because we expect this feature to be used to reduce thread use. If the goal is to run work on another thread, then return the result, there's no reason to return a future, one can execute that work synchronously in the service method.

In some cases it may be useful for specific types of endpoints to use executors to allow certain types of tasks to queue for execution on a bounded number of threads, but I don't think this framework is the correct place to manage that executor.

Reply:
I think this is quite easy to abuse / needs a lot of understanding of undertow internals and threading strategies. Thoughts on how to avoid bad "pseudo-async" uses? Maybe we should write 2-3 examples of "good" and "bad".

Reply:
> I think this is quite easy to abuse / needs a lot of understanding of undertow internals and threading strategies

The intent of the design was to avoid needing to understand how the server works behind the scenes, and provide something that works as expected. The thread model I described should only need to be understood by developers writing code for the undertow generator and runtime.

Could you describe how this could be abused? I think there are ways it could be used inefficiently, but that's true of existing generator implementations as well.

I think I may have created confusion in this test, the handler implementation uses hand-written handler code because I had not written the generator yet at this point. All uses of the async component should be done through code we generate, which provides little to no room for abuse. I'm planning to replace these tests with a set which use generated handler code instead.

> Thoughts on how to avoid bad "pseudo-async" uses? Maybe we should write 2-3 examples of "good" and "bad".

Good idea, do you think the readme is a good place for these, or perhaps a separate `docs/undertow_async.md` document? Alternatively I can embed examples in javadoc.

Reply:
I'd put it in the README. Naively, if I saw I should implement a
````
interface MyService {
  ListenableFuture<String> foo();
}
```
, I would make a `MyResource`, have it eat an ExecuterService, and then in my `MyServer` init method create something like a `Executors.newFixedThreadPool` with some arbitrary number, and then compute my futures on that thread pool. And then, I'd think, I'd have a perfectly async service with great performance. I imagine that's not what you want me to do :)

Reply:
Will do, thank you for the example :-)